doc/adrs/051_add_support_for_editing_context_action.adoc

```diff
@@ -0,0 +1,57 @@
+= ADR-051 - Add Support for Editing Context Action
+
+== Context
+
+Stereotypes descriptions are hard-coded and limited. The main thing that they can do is to copy JSON into the resource set. 
+We want to do more things, for example defining when an action is available based on the editing context.
+Replacing stereotypes descriptions would also reduce the dependency to the concept of Document.
+
+== Decision
+
+In order to allow for more flexibility, Stereotypes descriptions will be deleted and replaced by Editing Context Actions. 
+They should be supported by both the programmatic API and the GraphQL API.
+Every action will not be available all the time. Their availability will be computed and will depend on the editing context.
+The actions will be able to manipulate the editing context or contact a remote service.
+
+We will follow the example of the ADR-041 - Add the ability to contribute additional services to the EMFQueryService meaning that we will add two new interfaces :
+
+```
+public interface IEditingContextAction {
+  boolean isApplicable(IEditingContext editingContext);
+  
+  void apply(IEditingContext editingContext);
+}
+```
+
+```
+public interface IEditingContextActionProvider {
+  List<IEditingContextAction> getEditingContextAction(IEditingContext editingContext);
+}
+```
+
+I am not sure which Service will gather and call the actions, on the other end. Do you have an idea? In the case of emf query services, it was the EMFQueryService. 
```

Comment:
You will need an `InvokeEditingContextActionEventHandler` which will be called by the GraphQL data fetcher in order to execute an action selected by the end user. Something like this:

```
package org.eclipse.sirius.components.collaborative.handlers;

@Service
public class InvokeEditingContextActionEventHandler implements IEditingContextEventHandler {

  private final List<IEditingContextActionHandler> editingContextActionHandlers;

  public boolean canHandle(IEditingContext editingContext, IInput input) {
    return input instanceof InvokeEditingContextActionInput;
  }

  public void handle(One<IPayload> payloadSink, Many<ChangeDescription> changeDescriptionSink, IEditingContext editingContext, IInput input) {
    // Use the relevant editingContextActionHandler to handle the input
  }
}
```

This event handler will be used to call the appropriate action handler which will be provided by the same person who has contributed the action:

```
public interface IEditingContextActionHandler {
  boolean canHandle(IEditingContext editingContext, InvokeEditingContextActionInput);

  IPayload handle(IEditingContext editingContext, InvokeEditingContextActionInput);
}
```

On top of that, you need to both an input and a payload which will look like that:

```
public final class InvokeEditingContextActionInput implements IInput {
  private UUID id;
  private String actionId;
}
```

and

```
public class InvokeEditingContextActionSuccessPayload implements IPayload {
  private UUID id;
}
```